### DISCUSSION TITLE
ZFS hole

### DISCUSSION THREAD

**Discussion by zheliazkov** [openzfs/zfs#12307] [Q&A] [answered]
Hello,

I'll be glad to get information/help about the following issue:

1. I've pool with with multiple raid-z2 vdevs and one log device
2. I wanted to replace the log device and removed it from the pool
3. The status of the zpool is currently ok
4. Now I want to put another devices as a read and write cache but when I try dry-run (-n) it shows some hole, and the log device is not in the log "section" of the output

```
# zpool add -n stor log /dev/disk/by-id/nvme-Corsair_MP400_....
would update 'stor' to the following configuration:
        stor
          raidz2-0
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
          raidz2-1
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
          hole-2
          raidz2-3
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_....
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
            scsi-SHGST_HUH728080AL4200_...
          nvme-Corsair_MP400_...
```
The questions are:
1. Why this hole appears in the dry-run and not in the current status and what it is? Is it a problem?
2. Why the new dev is not in a log "section"?

Thanks in advance.

BR,
Plamen

**Comment by bghira** [comment]
i think this is where the temporary vdev removal indirection lives, in a virtual device called hole-_n_. i've never played with that before, so it's just a guess.

**Reply by bghira** [reply]
the reason it shows up in dry run but not otherwise is likely a bug (ditto for the lack of log indent). this code is very delicate, as are most string manipulations in life.

**Comment by AttilaFueloep** [comment, marked as answer]
All fixed in OpenZFS 2.0.2, see  #11442, #11167 and #11311.

**Reply by AttilaFueloep** [reply]
https://github.com/openzfs/zfs/issues/10889#issuecomment-803521902 is also related.
